### PR TITLE
Add local overrides to testbed and provide more defaults

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Allow providing overrides to a single test run and add more defaults to the overrides.

### CODE REVIEW

File: packages/flutter_tools/test/src/testbed.dart [hunk 1]
```diff
@@ -61,11 +80,19 @@ class Testbed {
...
+  ///
+  /// `overrides` may be used to provide new context values for the single test
+  /// case or override any context values from the setup.
+  FutureOr<T> run<T>(FutureOr<T> Function() test, {Map<Type, Generator> overrides}) {
     final Map<Type, Generator> testOverrides = Map<Type, Generator>.from(_testbedDefaults);
+    // Add the initial setUp overrides
     if (_overrides != null) {
       testOverrides.addAll(_overrides);
     }
+    // Add the test-specific overrides
```
Comment (zanderso) [inline packages/flutter_tools/test/src/testbed.dart]: Some tests that use 'is' checks to make sure that the overrides work as documented would be good. What will/should happen with different combos of overrides appearing {in the default list above, in the Testbed constructor, in the run call}?

Comment (jonahwilliams) [reply packages/flutter_tools/test/src/testbed.dart]: Added tests

File: packages/flutter_tools/test/src/testbed.dart [hunk 2]
```diff
@@ -7,15 +7,32 @@ import 'dart:async';
...
   FileSystem: () => MemoryFileSystem(),
   Logger: () => BufferLogger(),
+  DeviceManager: () => MockDeviceManager(),
+  Doctor: () => MockDoctor(),
+  FlutterVersion: () => MockFlutterVersion(),
+  HttpClient: () => MockHttpClient(),
+  OutputPreferences: () => OutputPreferences(showColor: false),
+  OperatingSystemUtils: () => MockOperatingSystemUtils(),
+  Usage: () => MockUsage(),
+  TimeoutConfiguration: () => const TimeoutConfiguration(),
```
Comment (zanderso) [inline packages/flutter_tools/test/src/testbed.dart]: Why is this needed?

File: packages/flutter_tools/test/src/testbed.dart [hunk 3]
```diff
@@ -7,15 +7,32 @@ import 'dart:async';
...
+import 'package:flutter_tools/src/version.dart';
+
+import 'context.dart';
 
 export 'package:flutter_tools/src/base/context.dart' show Generator;
 
 final Map<Type, Generator> _testbedDefaults = <Type, Generator>{
   FileSystem: () => MemoryFileSystem(),
   Logger: () => BufferLogger(),
+  DeviceManager: () => MockDeviceManager(),
```
Comment (zanderso) [inline packages/flutter_tools/test/src/testbed.dart]: This list needs some comments describing how its membership is decided. What is the procedure for deciding that a mock should be in this defaults list?

File: packages/flutter_tools/test/src/testbed.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/src/testbed.dart]: Removed, it is not needed

File: packages/flutter_tools/test/src/testbed.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/src/testbed.dart]: Added a standard criteria and removed those that did not fit. Also documented the reason for each current defaults inclusion

**Review by zanderso — APPROVED** [review]
